pkg/cloud/gcp/actuators/machine/actuator.go

```diff
@@ -96,7 +97,31 @@ func (a *Actuator) Exists(ctx context.Context, machine *machinev1.Machine) (bool
 	// When create()/update() try to store machineSpec/status this might result in
 	// "Operation cannot be fulfilled; the object has been modified; please apply your changes to the latest version and try again."
 	// Therefore we don't close the scope here and we only store spec/status atomically either in create()/update()"
-	return newReconciler(scope).exists()
+	exists, err := newReconciler(scope).exists()
+	if !isInvalidMachineConfigurationError(err) {
+		return exists, err
+	}
+
+	if exists {
+		// If the machine configuration becomes invalid, we have to Fail the machine.
+		// Worker machines don't become Failed automatically and they would be stuck
+		// in Running phase with invalid configuration.
+		machine.Status.Phase = pointer.String("Failed")
+	} else {
+		// If the machine has, e.g. invalid zone, and it doesn't have a phase set yet,
+		// we have to make sure the phase goes as "Failed".
+		if machine.Status.Phase == nil {
+			machine.Status.Phase = pointer.String("Failed")
+		}
+		// If the machine has, e.g. invalid zone, and we delete the invalid machinset,
+		// we want to set the machine to "Deleting" phase and return nil as error.
+		// We need the error to be nil so we can successfully delete the invalid machine.
+		if *machine.Status.Phase == "Deleting" {
+			return false, nil
+		}
+	}
+
+	return exists, err
```

Comment:
So, I think we can refactor this a little to simplify it. One thing I'm uncomfortable is changing the meaning of the `exists` bool within `exists()`, so I think we should do the providerID check closer to where that information is being used, I think it makes it easier to follow in the long term
```suggestion
	exists, err := newReconciler(scope).exists()
	if !isInvalidMachineConfigurationError(err) {
		return exists, err
	}

	// If we got here, then the machine configuration is invalid.

	// If the machine has, e.g. invalid zone, and it has no phase set,
	// we want to set the machine to "Failed" phase so that we don't attempt
	// to create it again.
	if machine.Status.Phase == nil {
		machine.Status.Phase = pointer.String("Failed")
	}

	// If the machine has, e.g. invalid zone, and we delete the invalid machinset,
	// we want to set the machine to "Deleting" phase and return nil as error.
	// We need the error to be nil so we can successfully delete the invalid machine.
	// If the machine has a provider ID, then we believe something exists in the cloud.
	// So we must not all the deletion if the provider ID is set.
	if *machine.Status.Phase == "Deleting" && machine.Spec.ProviderID == nil {
		return false, nil
	}

	return false, err
```

Breaking this down, this first part retains the old behaviour and then explains that for the rest of this block the machine config is invalid
```
	exists, err := newReconciler(scope).exists()
	if !isInvalidMachineConfigurationError(err) {
		return exists, err
	}

	// If we got here, then the machine configuration is invalid.
```
Then, if the phase is nil, this means this is the first time we are seeing this Machine and it's been invalid from the start, so we can move it to failed. If the phase is nil, it definitely won't exist (unless something really bad has happened which is out of scope), so we don't need any other checks wrapping this.
```
	// If the machine has, e.g. invalid zone, and it has no phase set,
	// we want to set the machine to "Failed" phase so that we don't attempt
	// to create it again.
	if machine.Status.Phase == nil {
		machine.Status.Phase = pointer.String("Failed")
	}
```
Then, we have the caveat of is the Machine deleting. We override the error when the Machine is in the deleting phase and we don't see a providerID. I think it's important to do the providerID check here as it makes it very clear where and why we are doing this, with the additional comments as well.
```
	// If the machine has, e.g. invalid zone, and we delete the invalid machinset,
	// we want to set the machine to "Deleting" phase and return nil as error.
	// We need the error to be nil so we can successfully delete the invalid machine.
	// If the machine has a provider ID, then we believe something exists in the cloud.
	// So we must not all the deletion if the provider ID is set.
	if *machine.Status.Phase == "Deleting" && machine.Spec.ProviderID == nil {
		return false, nil
	}
```
Finally, return a false explicitly, we should always return false if we don't know, it's up to the caller to handle the error.
```
	return false, err
```

With this change, we then need a slight change to the other file too, I'll make suggestions for those as well

Reply:
I think you are forgetting about machine that used to be valid and is now invalid. This case is not covered and such machine would be stuck in `running` even though it should be failed I think.